### ISSUE TITLE
Memory leak in concurrent.join

### ISSUE THREAD

**Issue by mpilquist** [functional-streams-for-scala/fs2#661]
Originally spotted by @amir and confirmed by @fiadliel. I minimized to the following program, which shows a very quick leak:

```scala
object ConcurrentJoinSanityTest extends App {
  import TestUtil.S
  concurrent.join(5)(Stream.constant(Stream.empty).covary[Task]).run.unsafeRun
}
```

**Comment by pchiusano** [comment]
Some random thoughts that might not be that helpful or which you probably have already thought about @mpilquist. 

From [@fiadliel's comment here](https://gitter.im/functional-streams-for-scala/fs2?at=576c1840c417e36c69e61eeb) sounds like resources are being allocated but never cleared. Might be the resources allocated by async steps, or some of the resources allocated as internal implementation details of `concurrent.join`.

* If you just do a regular join - `Stream.constant(Stream.empty).flatMap(identity).covary[Task].run.unsafeRun`, that is okay?
* Is it something with the empty stream? Like if you do `Stream.constant(Stream(1))`, is the leak still there?
* I was trying to think of some way of isolating whether it is the resources allocated by the async steps or implementation details of `concurrent.join`. What you could do is create a class `Foo` and make sure all the resources allocated by async steps keep a reference to a `Foo` (which they ignore). Then in the profiler you can see if the number of reachable `Foo` objects on the heap is ever much greater than 5.

**Comment by mpilquist** [comment]
> If you just do a regular join - ..., that is okay?

Yep
> Is it something with the empty stream? Like if you do Stream.constant(Stream(1)), is the leak still there?

Nothing specific about the empty stream. Simple pure streams or effectful streams also reproduce the issue.

> I was trying to think of some way of isolating whether it is the resources allocated by the async steps or implementation details of concurrent.join.

I think we'd be able to repeat this with a simple merge if the bug was in async steps. For example, this program runs in constant memory:
```scala
  val x = Stream.repeatEval(Task.delay(())).take(1000000)
  (x merge x).run.unsafeRun
```

This one fills the heap, but I suspect it is still constant memory usage -- just a very large constant:
```scala
  val x = Stream.empty.covary[Task]
  List.fill(1000000)(x).foldLeft(x)(_ merge _).run.unsafeRun
```

**Comment by mpilquist** [comment]
Just found something interesting...

I've been meaning to add `joinUnbounded` -- a copy/paste of `join` but with the `maxOpen` parameter removed. Use case is merging a small set of streams where you don't care about limiting the open streams, but you also don't want to pay the overhead of tracking open streams (e.g., the overhead that occurs when you set `maxOpen = Int.MaxValue` or whatever).

So I copy/pasted `join` and removed the `throttle` function, leaving this:

```scala
def joinUnbounded[F[_],O](outer: Stream[F,Stream[F,O]])(implicit F: Async[F]): Stream[F,O] = {

    for {
      killSignal <- Stream.eval(async.signalOf(false))
      outputQueue <- Stream.eval(async.mutable.Queue.synchronousNoneTerminated[F,Either[Throwable,Chunk[O]]])
      o <- outer.map { inner =>
        inner.chunks.attempt.evalMap { o => outputQueue.enqueue1(Some(o)) }.interruptWhen(killSignal)
      }.onFinalize {
        outputQueue.enqueue1(None)
      }.mergeDrainL {
        outputQueue.dequeue.through(pipe.unNoneTerminate).flatMap {
          case Left(e) => Stream.eval(killSignal.set(true)).flatMap { _ => Stream.fail(e) }
          case Right(c) => Stream.chunk(c)
        }
      }.onFinalize { killSignal.set(true) }
    } yield o
  }
```

Then I ran the following program, which *did not* leak:

```scala
  concurrent.joinUnbounded(Stream.constant(Stream.empty).covary[Task]).run.unsafeRun
```

So whatever the issue is, it's very likely to be caused by something in the implementation of `throttle`.

**Comment by pchiusano** [comment]
The plot thickens... seems like good progress. :)

**Comment by mpilquist** [comment]
@pchiusano So, by design, we defer cleanup of all resources acquiring during a pull until the pull is closed, right? If so, `throttle` will never work.

**Comment by mpilquist** [comment]
Fixed in #662
